README.md
README
================

Install the newest versions of the CTRM, MittagLeffleR and rticles
packages to get the paper (`article_springer.Rmd`) to compile:

``` r
devtools::install_github("strakaps/CTRM")
devtools::install_github("strakaps/MittagLeffleR")
devtools::install_github("rstudio/rticles")
library(CTRM)
library(MittagLeffleR)
library(rticles)
```


changes.md
# Response to the reviewers’ comments on the paper

We would like to thank the reviewers very much for their careful reviews and the helpful comments that gave us the opportunity to improve the manuscript.
Below we explain point by point how we addressed the comments during our revision of the manuscript.


## Referee 1

### Major points and questions

> It should be noted, however, that this process has light-tailed
Mittag-Leffler marginals, while the authors of the paper under review encounter a different, heavy-tailed Mittag-Leffler law.

+ We have added the following sentence near the end of Section 3: 
"We caution the reader that, somewhat confusingly, there is another distribution
called the "light-tailed" Mittag-Leffler distribution.  This is in fact the 
limiting distribution of the renewal process $N(t)$ above (see @limitCTRW)."
+ We have also added the following reference: 
Meerschaert, Mark M, and Hans-Peter Scheffler. 2004. “Limit Theorems forContinuous-Time Random Walks with Infinite Mean Waiting Times.”J. Appl.Probab.41 (3): 623–38. https://doi.org/10.1239/jap/1091543414.

> (p.5, l. 9-13) Why is it true that the limiting distribution exists? It is well known... 

In case of a continuous distribution one can choose the normalization in such a way that a non-degenerate limit distributions exists, but this is not neccessarily true for a linear normalization. We just forgot to add the assumption, that the distribution of the magnitudes is out of the domain of attraction of an extreme value distribution. We added this accordingly. 

> Note: It is important to observe that due to the fact that the waiting times Wj ’s have infinite
means, the renewal counting process N(t) cannot be made stationary (through a random shift) and hence this model does not fall into the classic context of Hsing-Leadbetter and Davis-Hsing.

We have added a remark at the end of Section 3, saying: 

"When $0 < \beta < 1$, the renewal process $N(t)$ is _not stationary_, 
and hence the results by @Hsing88 on the exceedances of stationary sequences
do not apply."

> Given the first comment above, can you elucidate the connection between the two different kinds of Mittag-Leffler distributions – the light-tailed distributional limit for the counting process N(t) and the heavy-tailed limit model for the T(l)’s?

We have addressed this, see first point. 

> The use of k is very confusing ... These two sentences are so unclear, so poorly written...  Be specific, choose your words carefully and introduce labels/titles in the plots.

We have re-written the entire section titled "Simulation study". 

> (Section 5 and 6) In these 2 sections it is implied that p = k/n .... While a complete rigor in the discussion on the statistical methodology is perhaps not necessary, itis important for the reader to clearly see where “asymptotic approximations” are applied.

We have re-written Sections 5 & 6, and made sure that $p$ is used as the theoretical exceedance probability, whereas the fraction of exceedances $k/n$ is denoted as $\hat p$. We have also clarified throughout where we assuming approximations.

> (Section 5) The methodology in this section (pages 9 and 10) is not very clearly described. Why give 2 algorithms if you are using just the second one, mostly? ... The scaling/rescaling of the scale coefficient by the choice of l and k is very confusing. Please revise Section 5 thoroughly.

We have re-written Section 5, as per the referee's suggestions, and we can now do without the "practical adjustments" section. 

> (page 11, line 12) Poor scientific language: “... the hill estimator gets useless ... Also a few lines after that, they claim that the scale cannot be estimated. This is not entirely true. ... 

KATHI TO DO

> Last but not least, given that the authors are dealing with more or less standard semi- parametric estimation of heavy-tail exponents and scales, they could formulate limit the- orems on the asymptotic distribution of their statistics and derive Wold-type confidence intervals for the estimated parameters. ... 

In the last paragraph of Section 5, we have included pointers to the literature for confidence intervals for the shape and scale parameters of the Mittag-Leffler distribution, and how these lead to "confidence bands" in our plots, based on the CTRE package. 


### Typos and suggestions

1. fixed
2. fixed
3. fixed
4. fixed
5. fixed
6. fixed
7. fixed
8. fixed
9. fixed
10. fixed
11. fixed
12. fixed
13. fixed
14. fixed
15. We reformulated the whole part and added titles. 
16. fixed
17. no longer applies
18. fixed
19. fixed
20. fixed
21. fixed
22. fixed
23. fixed
24. fixed
25. fixed
26. The section has been re-written
27. fixed
28. fixed
29. fixed
30. fixed
31. fixed
32. fixed
33. We have commented as follows: 
    "The stability plot for the tail stabilizes nicely around 0.85 (dashed line), while 
    the scale parameter stabilizes less obviously near $3 \times 10^7$ (dashed line).
    The growth of the scale parameter for lower threshold appears to be closer to linear in $p$, 
    rather than proportional to $p^{1/0.85}$ as suggested by the Mittag-Leffler
    fits. The reason for this is likely that the overall goodness of fit as 
    compared to an exponential distribution is improved due to the 
    peaked shape of the Mittag-Leffler distribution near $0$,
    rather than its tail behaviour at $\infty$.
    The reported fit should hence come with the caveat that a Mittag-Leffler 
    distribution models exceedance times well only up to certain time-scales. 
    More research is needed into the modelling of scale transitions, where
    inter-exceedance times appear to have different power laws across different 
    time scales."
34. We have changed "fitted" to "estimated".
35. We have added confidence limits.
36. We have added titles to the two figures, and to other figures where 
    feasible. 
37. fixed
38. fixed

## Referee 2


article_springer/article_springer_files/ML_estim_simu/ML_estim_simustudy.R
#Simulation Study #####

library(future)
library(future.apply)
library(MittagLeffleR)
library(dplyr)
library(TLMoments)
library(progress)
library(magrittr)
library(ggplot2)


plan(multiprocess)

tails <-0.2# c(0.2,0.8) #seq(0.2,0.9,0.1)
scales <- 1 #c(0.1,1,10)
ns <-25# seq(25,200,25)
m<-1000

pb <- progress_bar$new(total = length(tails)*length(scales)*length(ns), format = "[:bar] tail: :tail, scale: :scale, n: :n")

estimates <- sapply(ns, function(n) sapply(tails, function(tail) sapply(scales, function(scale) {
  pb$tick(tokens = list(tail = tail, scale = scale, n = n))
  future_replicate(m, {


    x <- rml(n, tail = tail, scale = scale)
    cbind(
      logMom = MittagLeffleR::logMomentEstimator(x)[c(1,2)],
      ML = MittagLeffleR::mlmle(x)$par
    )

  })
}, simplify = "array"), simplify = "array"), simplify = "array")


reshape2::melt(estimates) %>%
  as.tbl %>%
  rename(param = Var1, method = Var2, rep = Var3, true.scale = Var4, true.tail = Var5, n = Var6) %>%
  mutate(true.scale = scales[true.scale], true.tail = tails[true.tail], n = ns[n]) %>%
  mutate(true.value = case_when(param == "tail" ~ true.tail, param == "scale" ~ true.scale)) %>%
  group_by(param, n, true.scale, true.tail, method) %>%
  summarise(
    mse = mean((value-true.value)^2),
    sqbias = mean(value-true.value)^2,
    var = var(value)
  ) ->
  tbl_ML_estim

save(tbl_ML_estim,file="tbl_ML_estim.RData")


article_springer/article_springer_files/Simulation_plots/Plots_SimuStudy.R
library(ggplot2)
library(gridExtra)
library(here)

path<-here("article_springer/article_springer_files/Simulation_results")
files<-list.files(path)
file_dirs<-paste(path,"/",files,sep="")
lapply(file_dirs,FUN=load,.GlobalEnv)

#Tail Parameter

# Example One

#logMoment estimator
logMom_Ex1_tail<-ggplot()+ geom_line(aes(x=k,y=value,group=rep),colour="lightgrey",data=tbl_estimates_tail08_logMoms_Ex1) +
  coord_cartesian(ylim = c(0.5, 1.5)) +
  geom_line(aes(x=k, y=mw),colour="black",data=means.tbl_logMoms_Ex1) + ylab("Estimated tail") + geom_hline(yintercept=0.8, linetype="dashed", color = "red")

#ml estimator
mle_Ex1_tail<-ggplot() + geom_line(aes(x=k,y=value,group=rep),colour="lightgrey",data=tbl_estimates_tail08_mles_Ex1) +
  coord_cartesian(ylim = c(0.5, 1.5)) +
  geom_line(aes(x=k, y=mw),colour="black",data=means.tbl_mles_Ex1) + ylab("Estimated tail") + geom_hline(yintercept=0.8, linetype="dashed", color = "red")

#Hill estimator
hill_Ex1_tail<-ggplot()+ geom_line(data=tbl_estimates_tail08_hill_Ex1,aes(x=k,y=value,group=rep),colour="lightgrey") +
  coord_cartesian(ylim = c(0.5, 1.5)) +
  geom_line(data=means.tbl_hill_Ex1,aes(x=k, y=mw),colour="black") + ylab("Estimated tail") +  geom_hline(yintercept=0.8, linetype="dashed", color = "red")


#Example Two

#logMoment estimator
logMom_Ex2_tail<-ggplot()+ geom_line(aes(x=k,y=value,group=rep),colour="lightgrey",data=tbl_estimates_Ex2_Pareto_tail08_logMoms) +
  coord_cartesian(ylim = c(0.5, 1.5)) +
  geom_line(aes(x=k, y=mw),colour="black",data=means.tbl_logMoms_Ex2) + ylab("Estimated tail") + geom_hline(yintercept=0.8, linetype="dashed", color = "red")

#ml estimator
mle_Ex2_tail<-ggplot() + geom_line(aes(x=k,y=value,group=rep),colour="lightgrey",data=tbl_estimates_Ex2_Pareto_tail08_mles) +
  coord_cartesian(ylim = c(0.5, 1.5)) +
  geom_line(aes(x=k, y=mw),colour="black",data=means.tbl_mles_Ex2) + ylab("Estimated tail") + geom_hline(yintercept=0.8, linetype="dashed", color = "red")

#Hill estimator
hill_Ex2_tail<-ggplot() + geom_line(aes(x=k,y=value,group=rep),colour="lightgrey",data=tbl_estimates_Ex2_Pareto_tail08_hill_times) +
  coord_cartesian(ylim = c(0.5, 1.5)) +
  geom_line(aes(x=k, y=mw),colour="black",data=means.tbl_hill_times_Ex2) + ylab("Estimated tail") +  geom_hline(yintercept=0.8, linetype="dashed", color = "red")

#Example Three

#logMoment estimator
logMom_Ex3_tail<-ggplot()+ geom_line(aes(x=k,y=value,group=rep),colour="lightgrey",data=tbl_estimates_Example3_Exp_logMoms) +
  coord_cartesian(ylim = c(0.5, 1.5)) +
  geom_line(aes(x=k, y=mw),colour="black",data=means.tbl_logMoms_Ex3) + ylab("Estimated tail") + geom_hline(yintercept=1, linetype="dashed", color = "red")

#ml estimator
mle_Ex3_tail<-ggplot() + geom_line(aes(x=k,y=value,group=rep),colour="lightgrey",data=tbl_estimates_Example3_Exp_mles) +
  coord_cartesian(ylim = c(0.5, 1.5)) +
  geom_line(aes(x=k, y=mw),colour="black",data=means.tbl_mles_Ex3) + ylab("Estimated tail") + geom_hline(yintercept=1, linetype="dashed", color = "red")

#Hill estimator
hill_Ex3_tail<-ggplot() + geom_line(aes(x=k,y=value,group=rep),colour="lightgrey",data=tbl_estimates_Example3_Exp_hill_times) +
  coord_cartesian(ylim = c(0.5, 10)) +
  geom_line(aes(x=k, y=mw),colour="black",data=means.tbl_hill_times_Ex3) + ylab("Estimated tail") + geom_hline(yintercept=1, linetype="dashed", color = "red")

grid.arrange(logMom_Ex1_tail,mle_Ex1_tail,hill_Ex1_tail,logMom_Ex2_tail,mle_Ex2_tail,hill_Ex2_tail,logMom_Ex3_tail,mle_Ex3_tail,hill_Ex3_tail,ncol=3)

#Scale Parameter

#Example One
logMom_Ex1_scale<-ggplot()+ geom_line(aes(x=k,y=value,group=rep),colour="lightgrey",data=tbl_scales_logMoms_Ex1) +
  coord_cartesian(ylim = c(0, 2)) +
  geom_line(aes(x=k, y=mw),colour="black",means.tbl_scales_logMoms_Ex1) + ylab("Estimated scale") + geom_hline(yintercept=1, linetype="dashed", color = "red")

mle_Ex1_scale<-ggplot()+ geom_line(aes(x=k,y=value,group=rep),colour="lightgrey",data=tbl_scales_mles_Ex1) +
  coord_cartesian(ylim = c(0, 2)) +
  geom_line(aes(x=k, y=mw),colour="black",data=means.tbl_scales_mles_Ex1) + ylab("Estimated scale") + geom_hline(yintercept=1, linetype="dashed", color = "red")

#Example Two

logMom_Ex2_scale<-ggplot()+ geom_line(aes(x=k,y=value,group=rep),colour="lightgrey",data=tbl_scales_logMoms_Ex2) +
  coord_cartesian(ylim = c(0, 2)) +
  geom_line(aes(x=k, y=mw),colour="black",means.tbl_scales_logMoms_Ex2) + ylab("Estimated scale") + geom_hline(yintercept=1, linetype="dashed", color = "red")

mle_Ex2_scale<-ggplot()+ geom_line(aes(x=k,y=value,group=rep),colour="lightgrey",data=tbl_scales_mles_Ex2) +
  coord_cartesian(ylim = c(0, 2)) +
  geom_line(aes(x=k, y=mw),colour="black",data=means.tbl_scales_mles_Ex2) + ylab("Estimated scale") + geom_hline(yintercept=1, linetype="dashed", color = "red")

#Example Three

logMom_Ex3_scale<-ggplot()+ geom_line(aes(x=k,y=value,group=rep),colour="lightgrey",data=tbl_scales_logMoms_Ex3) +
  coord_cartesian(ylim = c(0, 2)) +
  geom_line(aes(x=k, y=mw),colour="black",means.tbl_scales_logMoms_Ex3) + ylab("Estimated scale") + geom_hline(yintercept=1, linetype="dashed", color = "red")

mle_Ex3_scale<-ggplot()+ geom_line(aes(x=k,y=value,group=rep),colour="lightgrey",data=tbl_scales_mles_Ex3) +
  coord_cartesian(ylim = c(0, 2)) +
  geom_line(aes(x=k, y=mw),colour="black",data=means.tbl_scales_mles_Ex3) + ylab("Estimated scale") + geom_hline(yintercept=1, linetype="dashed", color = "red")

grid.arrange(logMom_Ex1_scale,mle_Ex1_scale,logMom_Ex2_scale,mle_Ex2_scale,logMom_Ex3_scale,mle_Ex3_scale,ncol=2)


article_springer/article_springer_files/Simulation_study_Rcode/Example1_burstPotPaper_scaletail08_Cluster.R
#Simulationsstudy - Comparisions of Inference procedure and different estimation methods

# Author: Katharina Hees
# Created: 03/12/2018

library("CTRE")
library("MittagLeffleR")
library("magrittr")
library("dplyr")
library("ggplot2")
library("parallel")


p <- as.integer(Sys.getenv("PBS_ARRAYID"))
print(p)

###Functions########################################

hill_estim<-function(k,data){
  n<-length(data)
  data<-sort(data)
  data_thinned<-data[(n-k):n]
  r<-1/k * sum(log(data_thinned/data_thinned[1]))
  return(1/r)
}

####################################################

#Example One  - CTRE with stable distributed waiting times - tail=0.8

n<-10000
tail<-0.8
sigma <- (cos(pi*tail/2))^(1/tail)

times<-stabledist::rstable(n,tail, 1, sigma, pm=1)
times_sum <- cumsum(times)
magnitudes <- extRemes::revd(n, scale = 1, shape = 0)
sim_ctre <- ctre(data.frame(times_sum, magnitudes))

ntasks <- 4
print(ntasks)

# Initialize Cluster
cl <- makeCluster(ntasks)
clusterExport(cl, "sim_ctre")
clusterExport(cl, "hill_estim")
clusterEvalQ(cl,library(MittagLeffleR))
clusterEvalQ(cl,library(CTRE))
clusterEvalQ(cl,library("magrittr"))

# Zufallszahlengeneratoren auf allen Workern des Clusters mit einem bestimmten
# seed initialisieren (verhindert Zyklen in den Zufallszahlen zwischen den
# Workern)
clusterSetRNGStream(cl,iseed=42)

#Code
ks<-c(30:500)
estimates_scaletail08<-parSapply(cl,ks, FUN = function(k){
  ias<-sim_ctre %>% thin(k=k) %>% interarrival
  cbind(
    logMoms=logMomentEstimator(ias)[c(1,2)],
    mles=mlmle(ias)$par[c(1,2)],
    hill_times=c(hill_estim(k=k,data=times),NA)
  )},simplify = "array")

stopCluster(cl = cl)


save(estimates_scaletail08, file=here("article_springer","article_springer_files","Simulation_results",paste("estimates_scaletail08_", p, ".RData", sep="")))

#############################################################################################################
#############################################################################################################
#write simulation results in one tibble

n<-10000
ks<-30:500
for (i in 1:100){
  load(paste0("estimates_scaletail08_",i,".RData"))
  estimates_scaletail08[2,,]<-sapply(ks,function(k){estimates_scaletail08[2,,k-29]*(k/n)^{1/0.8}})
  estimates_scaletail08<-reshape2::melt(estimates_scaletail08) %>% as.tbl %>% rename(param = Var1 ,method = Var2, k=Var3) %>% mutate(k = ks[k]) %>% group_by(k,param,method)
  estimates_scaletail08<-mutate(estimates_scaletail08,rep=i)
  assign(paste0("estimates_scaletail08_",i),estimates_scaletail08)
}

tbl_estimates_scaletail08<-estimates_scaletail08_1
for(i in 2:100){
  estimates_scaletail08<-get(paste0("estimates_scaletail08_",i))
  tbl_estimates_scaletail08<-bind_rows(tbl_estimates_scaletail08,estimates_scaletail08)
}

save(estimates_scaletail08, file=here("article_springer","article_springer_files","Simulation_results",paste("estimates_Example1_stable_scaletail08_", p, ".RData", sep="")))


# Tail parameter
#logMoms
tbl_estimates_scaletail08 %>% filter(param=="tail") %>%  filter(method=="logMoms") -> tbl_estimates_tail08_logMoms_Ex1
save(tbl_estimates_tail08_logMoms_Ex1, file=here("article_springer","article_springer_files","Simulation_results","tbl_estimates_tail08_logMoms_Ex1.RData"))
tbl_estimates_tail08_logMoms_Ex1 %>% summarise(mw = mean(value)) -> means.tbl_logMoms_Ex1
save(means.tbl_logMoms_Ex1, file=here("article_springer","article_springer_files","Simulation_results","means.tbl_logMoms_Ex1.RData"))

#mles
tbl_estimates_scaletail08 %>% filter(param=="tail") %>%  filter(method=="mles") -> tbl_estimates_tail08_mles_Ex1
save(tbl_estimates_tail08_mles_Ex1, file=here("article_springer","article_springer_files","Simulation_results","tbl_estimates_tail08_mles_Ex1.RData"))
tbl_estimates_tail08_mles_Ex1 %>% summarise(mw = mean(value)) -> means.tbl_mles_Ex1
save(means.tbl_mles_Ex1, file=here("article_springer","article_springer_files","Simulation_results","means.tbl_mles_Ex1.RData"))


#hill
load(here("article_springer","article_springer_files","Simulation_results","hill_estim.RData"))
tbl_estimates_tail08_hill_Ex1<-reshape2::melt(estimates_tail08_hill) %>% as.tbl %>% rename(k = Var1, rep= Var2) %>% mutate(k = ks[k]) %>% group_by(k)
tbl_estimates_tail08_hill_Ex1 %>% summarise(mw = mean(value))->means.tbl_hill_Ex1
save(tbl_estimates_tail08_hill_Ex1,file=here("article_springer","article_springer_files","Simulation_results","tbl_estimates_tail08_hill_Ex1.RData"))
save(means.tbl_hill_Ex1,file=here("article_springer","article_springer_files","Simulation_results","means.tbl_hill_Ex1.RData"))

#Scale Parameter

#logMoms
tbl_estimates_scaletail08 %>% filter(param=="scale") %>% filter(method=="logMoms") -> tbl_scales_logMoms_Ex1
means.tbl %>% filter(param=="scale") %>% filter(method=="logMoms") -> means.tbl_scales_logMoms_Ex1
save(tbl_scales_logMoms_Ex1,file=here("article_springer","article_springer_files","Simulation_results","tbl_scales_logMoms_Ex1.RData"))
save(means.tbl_scales_logMoms_Ex1,file=here("article_springer","article_springer_files","Simulation_results","means.tbl_scales_logMoms_Ex1.RData"))

#mles
tbl_estimates_scaletail08 %>% filter(param=="scale") %>% filter(method=="mles") -> tbl_scales_mles_Ex1
means.tbl %>% filter(param=="scale") %>% filter(method=="mles") -> means.tbl_scales_mles_Ex1
save(tbl_scales_mles_Ex1,file=here("article_springer","article_springer_files","Simulation_results","tbl_scales_mles_Ex1.RData"))
save(means.tbl_scales_mles_Ex1,file=here("article_springer","article_springer_files","Simulation_results","means.tbl_scales_mles_Ex1.RData"))


article_springer/article_springer_files/Simulation_study_Rcode/Example2_burstPotPaper_scaletail08_Cluster.R
#Simulationsstudy - Comparisions of Inference procedure and different estimation methods

# Author: Katharina Hees
# Created: 03/12/2018


library("CTRE")
library("MittagLeffleR")
library("magrittr")
library("dplyr")
library("ggplot2")
library("parallel")
library("ReIns")


p <- as.integer(Sys.getenv("PBS_ARRAYID"))
print(p)

###Functions########################################

hill_estim<-function(k,data){
  n<-length(data)
  data<-sort(data)
  data_thinned<-data[(n-k):n]
  r<-1/k * sum(log(data_thinned/data_thinned[1]))
  return(1/r)
}

###################################################

#Example One  - CTRE with Pareto distributed waiting times - tail=0.8
#1_1
n<-10000
tail<-0.8
C<-(1/gamma(1-tail))^{1/tail}
times<-cumsum(rpareto(n,shape=tail,scale=C))
magnitudes<-extRemes::revd(n, scale = 1, shape = 0)
sim_ctre<-ctre(data.frame(times, magnitudes))

ntasks <- 4
print(ntasks)

# Cluster initialisieren
cl <- makeCluster(ntasks)
clusterExport(cl, "sim_ctre")
clusterExport(cl, "hill_estim")
clusterEvalQ(cl,library(MittagLeffleR))
clusterEvalQ(cl,library(CTRE))
clusterEvalQ(cl,library("magrittr"))

# Zufallszahlengeneratoren auf allen Workern des Clusters mit einem bestimmten
# seed initialisieren (verhindert Zyklen in den Zufallszahlen zwischen den
# Workern)
clusterSetRNGStream(cl,iseed=42)

#Code
ks<-c(30:500)
estimates_scaletail08<-parSapply(cl,ks, FUN = function(k){
  ias<-sim_ctre %>% thin(k=k) %>% interarrival
  cbind(
    logMoms=logMomentEstimator(ias)[c(1,2)],
    mles=mlmle(ias)$par[c(1,2)],
    hill_times=c(hill_estim(k=k,data=times),NA)
  )},simplify = "array")

stopCluster(cl = cl)


save(estimates_scaletail08, file=here("article_springer","article_springer_files","Simulation_results",paste("estimates_Example2_Pareto_scaletail08_", p, ".RData", sep="")))


#############################################################################################################
#############################################################################################################
#write simulation results in one tibble

n<-10000
ks<-30:500
for (i in 1:100){
  load(paste0("estimates_scaletail08_",i,".RData"))
  estimates_scaletail08[2,,]<-sapply(ks,function(k){estimates_scaletail08[2,,k-29]*(k/n)^{1/0.8}/(1/gamma(0.2))^{1/0.8}*1/(gamma(0.2))})
  estimates_scaletail08<-reshape2::melt(estimates_scaletail08) %>% as.tbl %>% rename(param = Var1 ,method = Var2, k=Var3) %>% mutate(k = ks[k]) %>% group_by(k,param,method)
  estimates_scaletail08<-mutate(estimates_scaletail08,rep=i)
  assign(paste0("estimates_scaletail08_",i),estimates_scaletail08)

}

tbl_estimates_scaletail08<-estimates_scaletail08_1
for(i in 2:100){
  estimates_scaletail08<-get(paste0("estimates_scaletail08_",i))
  tbl_estimates_scaletail08<-bind_rows(tbl_estimates_scaletail08,estimates_scaletail08)
}

saveRDS(tbl_estimates_scaletail08, file="tbl_estimates_scaletail08_Ex2.rds")



#Tail Parameter

#logMoms
tbl_estimates_scaletail08 %>% filter(param=="tail") %>% filter(method=="logMoms") -> tbl_estimates_Ex2_Pareto_tail08_logMoms
tbl_estimates_Ex2_Pareto_tail08_logMoms %>% summarise(mw = mean(value)) ->means.tbl_logMoms_Ex2

#mles
tbl_estimates_scaletail08 %>% filter(param=="tail") %>% filter(method=="mles") -> tbl_estimates_Ex2_Pareto_tail08_mles
tbl_estimates_Ex2_Pareto_tail08_mles %>% summarise(mw = mean(value)) -> means.tbl_mles_Ex2

#hill
tbl_estimates_scaletail08 %>% filter(param=="tail") %>% filter(method=="hill_times") -> tbl_estimates_Ex2_Pareto_tail08_hill_times
tbl_estimates_Ex2_Pareto_tail08_hill_times %>% summarise(mw = mean(value)) -> means.tbl_hill_times_Ex2

save(tbl_estimates_Ex2_Pareto_tail08_logMoms, file=here("article_springer","article_springer_files","Simulation_results","tbl_estimates_Ex2_Pareto_tail08_logMoms.RData"))
save(tbl_estimates_Ex2_Pareto_tail08_mles, file=here("article_springer","article_springer_files","Simulation_results","tbl_estimates_Ex2_Pareto_tail08_mles.RData"))
save(tbl_estimates_Ex2_Pareto_tail08_hill_times, file=here("article_springer","article_springer_files","Simulation_results","tbl_estimates_Ex2_Pareto_tail08_hill_times.RData"))
save(means.tbl_logMoms_Ex2, file=here("article_springer","article_springer_files","Simulation_results","means.tbl_logMoms_Ex2.RData"))
save(means.tbl_mles_Ex2, file=here("article_springer","article_springer_files","Simulation_results","means.tbl_mles_Ex2.RData"))
save(means.tbl_hill_times_Ex2, file=here("article_springer","article_springer_files","Simulation_results","means.tbl_hill_times_Ex2.RData"))

#Scale Parameter

#logMoms
tbl_estimates_scaletail08 %>% filter(param=="scale") %>% filter(method=="logMoms") -> tbl_scales_logMoms_Ex2
means.tbl %>% filter(param=="scale") %>% filter(method=="logMoms") -> means.tbl_scales_logMoms_Ex2
save(tbl_scales_logMoms_Ex2, file=here("article_springer","article_springer_files","Simulation_results","tbl_scales_logMoms_Ex2.RData"))
save(means.tbl_scales_logMoms_Ex2, file=here("article_springer","article_springer_files","Simulation_results","means.tbl_scales_logMoms_Ex2.RData"))

#mles
tbl_estimates_scaletail08 %>% filter(param=="scale") %>% filter(method=="mles") -> tbl_scales_mles_Ex2
means.tbl %>% filter(param=="scale") %>% filter(method=="mles") -> means.tbl_scales_mles_Ex2
save(tbl_scales_mles_Ex2, file=here("article_springer","article_springer_files","Simulation_results","tbl_scales_mles_Ex2.RData"))
save(means.tbl_scales_mles_Ex2, file=here("article_springer","article_springer_files","Simulation_results","means.tbl_scales_mles_Ex2.RData"))



article_springer/article_springer_files/Simulation_study_Rcode/Example3_burstPotPaper_scaletail08_Cluster.R
#Simulationsstudy - Comparisions of Inference procedure and different estimation methods

# Author: Katharina Hees
# Created: 03/12/2018


library("CTRE")
library("MittagLeffleR")
library("magrittr")
library("dplyr")
library("ggplot2")
library("ReIns")
library("parallel")


p <- as.integer(Sys.getenv("PBS_ARRAYID"))
print(p)

###Functions########################################

hill_estim<-function(k,data){
  n<-length(data)
  data<-sort(data)
  data_thinned<-data[(n-k):n]
  r<-1/k * sum(log(data_thinned/data_thinned[1]))
  return(1/r)
}

####################################################


#Example Three - CTRE with Exponential Waiting times

#Daten generieren

#3_1
n<-10000

times<-rexp(n,rate=1)
sum_times <- cumsum(times)
magnitudes <- extRemes::revd(n, scale = 1, shape = 0)
sim_ctre <- ctre(data.frame(sum_times, magnitudes))

ntasks <- 4
print(ntasks)


# Cluster initialisieren
cl <- makeCluster(ntasks)
clusterExport(cl, "sim_ctre")
clusterExport(cl, "times")
clusterExport(cl, "hill_estim")
clusterEvalQ(cl,library(MittagLeffleR))
clusterEvalQ(cl,library(CTRE))
clusterEvalQ(cl,library("magrittr"))

# Zufallszahlengeneratoren auf allen Workern des Clusters mit einem bestimmten
# seed initialisieren (verhindert Zyklen in den Zufallszahlen zwischen den
# Workern)
clusterSetRNGStream(cl,iseed=42)



#Code
ks<-30:500
estimates_Example3_Exp_scaletail08<-parSapply(cl,ks, FUN = function(k){
  ias<-sim_ctre %>% thin(k=k) %>% interarrival
  cbind(
    logMoms=logMomentEstimator(ias)[c(1,2)],
    mles=mlmle(ias)$par[c(1,2)],
    hill_times=c(hill_estim(k=k,data=times),NA)
  )},simplify = "array")


# Cluster beenden
stopCluster(cl)

save(estimates_Example3_Exp_scaletail08, file=paste("estimates_Example3_Exp_scaletail08_", p, ".RData", sep=""))

#############################################################################################################
#############################################################################################################
#write simulation results in one tibble

n<-10000
ks<-30:500
for (i in 1:100){
  load(paste0("estimates_Example3_Exp_scaletail08_",i,".RData"))
  estimates_Example3_Exp_scaletail08[2,,]<-sapply(ks,function(k){estimates_Example3_Exp_scaletail08[2,,k-29]*(k/n)^{1}})
  estimates_Example3_Exp_scaletail08<-reshape2::melt(estimates_Example3_Exp_scaletail08) %>% as.tbl %>% rename(param = Var1 ,method = Var2, k=Var3) %>% mutate(k = ks[k]) %>% group_by(k,param,method)
  estimates_Example3_Exp_scaletail08<-mutate(estimates_Example3_Exp_scaletail08,rep=i)
  assign(paste0("estimates_Example3_Exp_scaletail08_",i),estimates_Example3_Exp_scaletail08)
}

tbl_estimates_Example3_Exp_scaletail08<-estimates_Example3_Exp_scaletail08_1
for(i in 2:100){
  estimates_Example3_Exp_scaletail08<-get(paste0("estimates_Example3_Exp_scaletail08_",i))
  tbl_estimates_Example3_Exp_scaletail08<-bind_rows(tbl_estimates_Example3_Exp_scaletail08,estimates_Example3_Exp_scaletail08)
}

saveRDS(tbl_estimates_Example3_Exp_scaletail08, file="tbl_estimates_Example3_Exp_scaletail08.rds")

tbl_estimates_Example3_Exp_scaletail08 %>% summarise(mw = mean(value)) -> means.tbl

#Tail
ks<-30:500
for (i in 1:100){
  load(paste0("estimates_Example3_Exp_",i,".RData"))
  estimates_Example3_Exp<-reshape2::melt(estimates_Example3_Exp) %>% as.tbl %>% rename(param = Var1 ,method = Var2, k=Var3) %>% mutate(k = ks[k]) %>% group_by(k)
  estimates_Example3_Exp<-mutate(estimates_Example3_Exp,rep=i)
  assign(paste0("estimates_Example3_Exp_",i),estimates_Example3_Exp)
}

tbl_estimates_Example3_Exp<-estimates_Example3_Exp_1
for(i in 2:100){
  estimates_Example3_Exp<-get(paste0("estimates_Example3_Exp_",i))
  tbl_estimates_Example3_Exp<-bind_rows(tbl_estimates_Example3_Exp,estimates_Example3_Exp)
}


#Tail
#logMoms
tbl_estimates_Example3_Exp_scaletail08 %>% filter(param=="tail") %>% filter(method=="logMoms") -> tbl_estimates_Example3_Exp_logMoms
tbl_estimates_Example3_Exp_logMoms %>% summarise(mw = mean(value)) -> means.tbl_logMoms_Ex3

#mles
tbl_estimates_Example3_Exp_scaletail08 %>% filter(param=="tail") %>% filter(method=="mles") -> tbl_estimates_Example3_Exp_mles
tbl_estimates_Example3_Exp_mles %>% summarise(mw = mean(value)) -> means.tbl_mles_Ex3

#hill
tbl_estimates_Example3_Exp_scaletail08 %>% filter(param=="tail") %>% filter(method=="hill_times") -> tbl_estimates_Example3_Exp_hill_times
tbl_estimates_Example3_Exp_hill_times %>% summarise(mw = mean(value)) -> means.tbl_hill_times_Ex3

save(tbl_estimates_Example3_Exp_logMoms, file=here("article_springer","article_springer_files","Simulation_results","tbl_estimates_Example3_Exp_logMoms.RData"))
save(tbl_estimates_Example3_Exp_mles, file=here("article_springer","article_springer_files","Simulation_results","tbl_estimates_Example3_Exp_mles.RData"))
save(tbl_estimates_Example3_Exp_hill_times, file=here("article_springer","article_springer_files","Simulation_results","tbl_estimates_Example3_Exp_hill_times.RData"))
save(means.tbl_logMoms_Ex3, file=here("article_springer","article_springer_files","Simulation_results","means.tbl_logMoms_Ex3.RData"))
save(means.tbl_mles_Ex3, file=here("article_springer","article_springer_files","Simulation_results","means.tbl_mles_Ex3.RData"))
save(means.tbl_hill_times_Ex3, file="means.tbl_hill_times_Ex3.RData")

#Scale


#logMoms
tbl_estimates_Example3_Exp_scaletail08 %>% filter(param=="scale") %>% filter(method=="logMoms") -> tbl_scales_logMoms_Ex3
means.tbl %>% filter(param=="scale") %>% filter(method=="logMoms") -> means.tbl_scales_logMoms_Ex3
save(tbl_scales_logMoms_Ex3, file=here("article_springer","article_springer_files","Simulation_results","tbl_scales_logMoms_Ex3.RData"))
save(means.tbl_scales_logMoms_Ex3,file=here("article_springer","article_springer_files","Simulation_results","means.tbl_scales_logMoms_Ex3.RData"))

#mles
tbl_estimates_Example3_Exp_scaletail08 %>% filter(param=="scale") %>% filter(method=="mles") -> tbl_scales_mles_Ex3
means.tbl %>% filter(param=="scale") %>% filter(method=="mles") -> means.tbl_scales_mles_Ex3
save(tbl_scales_mles_Ex3, file=here("article_springer","article_springer_files","Simulation_results","tbl_scales_mles_Ex3.RData"))
save(means.tbl_scales_mles_Ex3, file=here("article_springer","article_springer_files","Simulation_results","means.tbl_scales_mles_Ex3.RData"))
